### PR TITLE
Fix FindPeersConnectedToPeer

### PR DESCRIPTION
**Pull request by jbrady42**
Updated handleFindPeer special case to return peer list.
Re-enabled tests for FindPeersConnectedToPeer.

### CODE REVIEW

File: routing/dht/handlers.go [hunk 1]
```diff
@@ -131,9 +131,10 @@ func (dht *IpfsDHT) handleFindPeer(ctx context.Context, p peer.ID, pmes *pb.Mess
 	resp := pb.NewMessage(pmes.GetType(), "", pmes.GetClusterLevel())
 	var closest []peer.ID
 
-	// if looking for self... special case where we send it on CloserPeers.
+	// if looking for self... special case where we are looking for connected peers.
+	// Used for FindPeersConnectedToPeer
 	if peer.ID(pmes.GetKey()) == dht.self {
-		closest = []peer.ID{dht.self}
+		closest = dht.peerstore.Peers()
```
Comment (whyrusleeping) [inline routing/dht/handlers.go]: not so sure about this. Might be best to not have this and just rely on closer peers to give us what we need.

**Comment by whyrusleeping** [comment]
So, in general we really dont want to override things, it gets confusing (like right now as im trying to figure out what that special case was for). So here: https://github.com/ipfs/go-ipfs/pull/1001/files#r27631064 we can probably just opt towards removing the special case entirely. I still need to spend some time groking the test to convince myself its right, it may take me a day or so to allocate that time.

File: routing/dht/handlers.go (hunk 1 above)
Comment (jbenet) [inline routing/dht/handlers.go]: this isn't right-- the `peerstore` of the dht may contain data about peers we're no longer connected to, and is also shared with other subsystems. 

maybe we can grab the peers we're connected to from the routing table:

```go
dht.routingTable.ListPeers()
```

thoughts @whyrusleeping ?

**Comment by whyrusleeping** [comment]
@jbrady42, I'd like to see what happens if we always just set closer to the closer peers, i.e.:
```go
closest = dht.betterPeersToQuery(pmes, p, CloserPeerCount)
```

Also, would you mind maybe writing one or two more tests to exercise this a bit? maybe a few different network configurations? I think this will work, and i'm glad we're working towards relay (Yay!!!), but I just want to be super careful not to break anything in the DHT.

**Comment by jbrady42** [comment]
Thanks for the feedback. @whyrusleeping, I tried out your suggesting above but did not get any results. It looks like peers are being filtered out in betterPeersToQuery because they are not closer than self. Another possible issue here is the number of connected peers returned is limited by the CloserPeerCount. Even with the filter disabled it only gets 4 results, so there’s a chance it is connected to the peer we are looking for, but just not reporting it. 

In the meantime, I’ll work on some more test network topologies and add them soon.

**Comment by whyrusleeping** [comment]
Ah, you are very right about that, we do filter out peers that arent closer... In that case, I think your solution makes sense, but as @jbenet points out you will need to use the routing tables peer list.

**Comment by whyrusleeping** [comment]
in response to the discussion on irc, the routing table will only have peers that we have interacted with through the DHT. The connect method you are using in the tests to connect the dhts together *should* be sending out pings, which registers the other peer in the routing table. If that does not seem to be the case, you may have found a bug.

**Comment by jbenet** [comment]
> the routing table will only have peers that we have interacted with through the DHT. 

When connecting to someone, the network triggers a notification, which calls `dht.Update(peer)` -- the timing on this can be tricky in tests -- given goroutine scheduler. So theoretically the dht should _consider_ all peers we connect to now. but yeah there may be a bug somewhere.

Btw @jbrady42 -- messing with the dht is pretty hard, mostly because our code needs to be way better. by all means go for it, but if your aim is to help (not necessarily attack the dht) may be easier to start elsewhere while you get used to the codebase.

**Comment by jbrady42** [comment]
Yeah I agree with you there. I’m not specifically going after the dht. My goal is to get the relay hooked in. I figured this was a good place to start since it can be used in a search for potential relay hosts, and was basically done already. 

Regarding the mismatch in peer count on the tests, I have done some testing with the number and it seems to work consistently up to about 30 nodes. After that it comes up a few short some of the time, but not every time. It does what it is supposed to basically, so if there is an issue I think it is elsewhere. For the test, I’m adding another net configuration, but think it's better to turn down the number of nodes to around 20 so it does not fail with a mismatch.

**Comment by whyrusleeping** [comment]
@jbrady42 thanks for working on this! the relay is something i've wanted to have in for a while and this will get us that much closer to it.

**Comment by whyrusleeping** [comment]
closing due to inactivity, please reopen as necessary

note: all pull requests older than three weeks may be closed in an effort to keep our open pull requests more focused.

**Pull request closed without merging**